mib-generator/cw_mib_generator.py

```diff
@@ -0,0 +1,145 @@
+Usage:
+  cw_mib_generator.py [--cwc-mib-dir=DIR]
+
+Options:
+  --cwc-mib-dir=DIR  Directory containing Clearwater Core MIB fragment
+"""
+
+import re
+import sys
+import os.path
+from string import Template
+from docopt import docopt
+
+# Project Clearwater MIB paths
+COMMON_MIB_PATH = "./CLEARWATER-MIB-COMMON"
+PC_EXTRAS_PATH = "./CLEARWATER-MIB-PC-EXTRAS"
+PC_MIB_OUTPUT_PATH = "../PROJECT-CLEARWATER-MIB"
+
+# Clearwater Core MIB names
+CWC_EXTRAS_NAME = "CLEARWATER-MIB-CWC-EXTRAS"
+CWC_MIB_OUTPUT_NAME = "METASWITCH-CLEARWATER-CORE-MIB"
+
+# Statement added to top of auto-generated MIBs
+EDIT_STATEMENT = "THIS MIB IS BUILT FROM A TEMPLATE - DO NOT EDIT DIRECTLY!"
+
+
+def print_err_and_exit(error_text):
+    # Prints an error message with the specified text, and exits with code 1.
+    sys.stderr.write("ERROR: {}\nFailed to generate MIB, exiting\n"
+                     .format(error_text))
+    sys.exit(1)
+
+
+def read_mib_fragment(mib_path):
+    # Try to read MIB fragment at specified path. Returns read data if
+    # successful.
+    try:
+        with open(mib_path, "r") as mib:
+            mib_data = mib.read()
+    except IOError as ioe:
+        print_err_and_exit("Could not read from MIB fragment:\n - {}"
+                           .format(ioe))
+
+    return mib_data
+
+
+def write_mib_file(mib_path, mib_contents):
+    # Try to write specified data to file at specified MIB path.
+    try:
+        with open(mib_path, "w") as mib_file:
+            mib_file.write(mib_contents)
+    except IOError as ioe:
+        print_err_and_exit("Could not write MIB file:\n - {}".format(ioe))
+
+
+def generate_title(mib_file_path):
+    # Try to extract MIB name from the file path, and append definitions
+    # statement.
+    try:
+        name = re.search(r"[\w\-]+$", mib_file_path).group(0)
+    except AttributeError:
+        print_err_and_exit("Could not find a valid MIB file name in: '{}'"
+                           .format(mib_file_path))
+
+    return name + " DEFINITIONS ::= BEGIN"
+
+
+def generate_mib(extras_file_path, output_mib_path):
+    # Read in the common and specific MIB fragments, fill in templated lines
+    # and write the complete MIB file.
+    full_common_mib_path = os.path.abspath(COMMON_MIB_PATH)
+    raw_common_data = read_mib_fragment(full_common_mib_path)
+    common_src_template = Template(raw_common_data)
+
+    substitute_dict = { 'title_statement' : generate_title(output_mib_path),
+                        'direct_edit_statement' : EDIT_STATEMENT }
+
+    try:
+        common_src = common_src_template.substitute(substitute_dict)
+    except ValueError as ve:
+        print_err_and_exit(
+            "Common MIB fragment contains an invalid placeholder:\n - {}"
+            .format(ve))
+    except KeyError as ke:
+        print_err_and_exit(
+            "Common MIB fragment contains unrecognised placeholder: {}"
+            .format(ke))
+
+    extras_src = read_mib_fragment(extras_file_path)
```

Comment:
Because there's no templating going on in the 'additional' fragments, only in the core fragment (specifically for the title statement, and to add a statement about direct editing).